src/java/org/apache/cassandra/utils/btree/BTree.java

```diff
@@ -1448,4 +1920,2355 @@ private static boolean isStopSentinel(long v)
+            if (newKeys + count <= MAX_KEYS)
+            {
+                System.arraycopy(buffer, 0, buffer, newKeys, count);
+                System.arraycopy(pred,   0, buffer, 0,       predSize);
+                                            buffer [predSize] = predNextKey;
+                count += newKeys;
+            }
+            else
+            {
+                if (savedBuffer == null)
+                    savedBuffer = new Object[MAX_KEYS];
+                System.arraycopy(pred, 0, savedBuffer, 0, predSize);
+                if (predSize == MAX_KEYS)
+                {
+                    savedNextKey = predNextKey;
+                }
+                else
+                {
+                    int removeKeys = MAX_KEYS - predSize;
+                    count -= removeKeys;
+                                                         savedBuffer [predSize] = predNextKey;
+                    System.arraycopy(buffer, 0,          savedBuffer, predSize + 1, MAX_KEYS - newKeys);
+                                                         savedNextKey      = buffer[MAX_KEYS - newKeys];
+                    System.arraycopy(buffer, removeKeys, buffer,      0,            count);
+                }
+            }
+        }
+
+        /**
+         * Invoked when we want to add a key to the leaf buffer, but it is full
+         */
+        void propagateOverflow()
+        {
+            // propagate the leaf we have saved in altLeafBuffer
+            // precondition: savedLeafCount == MAX_KEYS
+            if (allocated >= 0)
+                allocated += ObjectSizes.sizeOfReferenceArray(MAX_KEYS);
+            ensureParent().addChildAndNextKey(savedBuffer, MAX_KEYS, savedNextKey);
+            savedBuffer = null;
+            savedNextKey = null;
+        }
+
+        /**
+         * Construct a new leaf from the contents of {@link #buffer}, unless the contents have not changed
+         * from {@code unode}, in which case return {@code unode} to avoid allocating unnecessary objects.
+         */
+        Object[] drainAndPropagate(Object[] unode, int usz, BranchBuilder propagateTo)
+        {
+            Object[] leaf;
+            int sizeOfLeaf;
+            if (mustRedistribute())
+            {
+                // we have too few items, so spread the two buffers across two new nodes
+                leaf = redistributeOverflowAndDrain();
+                sizeOfLeaf = MIN_KEYS;
+            }
+            else if (!hasOverflow() && count == usz && areIdentical(buffer, 0, unode, 0, usz))
+            {
+                // we have exactly the same contents as the original node, so reuse it
+                leaf = unode;
+                sizeOfLeaf = usz;
+            }
+            else
+            {
+                // we have maybe one saved full buffer, and one buffer with sufficient contents to copy
+                if (hasOverflow())
+                    propagateOverflow();
+
+                sizeOfLeaf = count;
```

Comment:
It can be, I think. If this leaf is the root. The mustRedistribute() path above captures all other cases, no?

When we are building a new tree, and we are not the root and so have overflowed at least once, we always maintain one full node at each level we are constructing plus some additional non-empty buffer.

In the case where we are inserting into an existing tree, we always have at least MIN_KEYS input on the side we are inserting into, so we must have at least MIN_KEYS+1 in the output.

In the case where we are transforming a tree and may remove items, we scavenge elements to ensure we always have MIN_KEYS.

Reply:
Another way of saying this is that a precondition for entering this method is that either this node is the root, or there are at least `MIN_KEYS` in the buffer, or that there is a `savedBuffer` and at least one key in the buffer. 

This translates to an invariant that the buffer is always non-empty, that `savedBuffer` is only set when there is some overflow that must be inserted to the buffer, and that a parent is only allocated once at least MIN_KEYS have been inserted to the buffer.